File: Matlab_Codes/Secant.m
%Secant method
%Written by Yusuf Hani Almoadhen 202107475

clc
clear 
close all

syms x
f = 3*x-cos(x)-2;

x0=0.5;
x1=1;
xk=0;
tol=100;
Error=10^-6;
index=1;
i=0;
k=input("Enter the number of iterations: ");

K=[];
X=[];
X0=[];
X1=[];
F0=[];
F1=[];
Tol=[];

while i<k && tol>=Error
    K(index,1) = i;
    X0(index,1) = x0;
    X1(index,1) = x1;
    f0 = subs(f,x,x0);
    F0(index,1) = f0;
    f1 = subs(f,x,x1);
    F1(index,1) = f1;
    xk = (x0*f1-x1*f0)/(f1-f0);
    X(index,1) = xk;
    if i==0
        Tol(index,1) = "-";
    else
        tol = abs(xk-x1);
        Tol(index,1) = tol;
    end
    x0 = x1;
    x1 = xk;
    index = index+1;
    i = i+1;
end

fprintf('\n')
T = table(K,X0,X1,F0,F1,X,Tol);
disp(T)

if k>index
    fprintf('\n')
    disp("You have exceeded the number of iterations accroding to the specified Tolerance!!")
end

fprintf('\n')
disp("The root is " + num2str(double(xk)))

File: Python Codes/Secant.py
#Secant method
#Written by Yusuf Hani Almoadhen

import pandas as pd
from sympy import *

x = Symbol('x')

eq = x**3-5*x+1

x0 = 0.0
x1 = 1.0
Tol = 100
Error = 1e-6
i = 0
k = int(input("Enter the number of iterations: "))
i_col=[]
x0_col=[]
x1_col=[]
f0_col=[]
f1_col=[]
Tol_col=[]
while i<k and Tol>=Error:
    i_col.append(str(i).zfill(2))
    x0_col.append("{:.5g}".format(x0))
    x1_col.append("{:.5g}".format(x1))
    f0 = float(eq.subs({x:x0}))
    f0_col.append("{:.5g}".format(f0))
    f1 = float(eq.subs({x:x1}))
    f1_col.append("{:.5g}".format(f1))
    xk = (x1*f0-x0*f1)/(f0-f1)
    if i==0:
        Tol_col.append('-')
    else:
        Tol = abs(xk-x1)
        Tol_col.append('{:.5g}'.format(Tol))
    x0 = x1
    x1 = xk
    i += 1
table = pd.DataFrame({
    'i': i_col,
    'x n-2': x0_col,
    'f(x n-2)': f0_col,
    'x n-1': x1_col,
    'f(x n-1)': f1_col,
    'Tol': Tol_col
})
print('')
print("Formula:",eq,'\n')
print(table.to_markdown(index=False),'\n')
if k>i:
    print("You have exceeded the maximum number of iterations according to the specified Tolerance!!",'\n')
print("The root is","{:.5g}".format(float(xk)))
